### ISSUE TITLE
C API

### ISSUE THREAD

**Issue by VarLad** [david-vanderson/dvui#65]
I would love to use this in projects outside Zig as well.
Any idea if this project can provide a C API?

**Comment by david-vanderson** [comment]
Good idea!  I haven't thought about this before.  Just to make sure, are you thinking to compile dvui into a static library and call the functions from C?

I think this should work with some modifications (adding extern to various functions).  Give me some time and I'll try to make a prototype example.

**Comment by VarLad** [comment]
@david-vanderson Yeah, exactly that

**Comment by iacore** [comment]
@VarLad you might be better off using imgui with C. Zig has types like `?T` that is incompatible with C ABI.

**Comment by david-vanderson** [comment]
I also recommend dear imgui, but I suspect you already know about it.  In any case it will still be useful to see if exposing a C API can be done.

**Comment by iacore** [comment]
it can be done but it is difficult.

Every `error.X` need a fixed number to map to it.
Every enum variant need a fixed number to map to it.
`union(enum)`: not exist.
`?T`: not exist except optional pointer.

**Comment by david-vanderson** [comment]
I spent some time digging into this.  It seems like this should be possible, but is currently not well supported.

https://github.com/ziglang/zig/issues/9698 looks like the tracking issue for this in zig.  I was able to use `zig build-lib -femit-h=test.h test.zig` to autogenerate a header for exported functions, but then ran into https://github.com/ziglang/zig/issues/14381.  All my attempts via build.zig also failed, sometimes with an error like `error: unable to create compilation: NoZigModuleForCHeader`

And this is all before getting into the issues that @iacore listed.

@VarLad It looks like this isn't the right time to attempt this, do you agree?  Maybe we should revisit this after zig 0.12 is released and see if things have gotten better.

**Comment by VarLad** [comment]
I agree with that. We can revisit the issue after Zig 0.12 release.

**Comment by Renha** [comment]
The tracking issue has 0.13.0 milestone set

**Comment by david-vanderson** [comment]
Good point.  If 0.12 doesn't fix it, then we might have to try doing it "manually" - using `-femit-h` but hand editing the output.

**Comment by VisenDev** [comment]
I did a little bit of testing related to this in C. It does appear to be possible to recreate zig's `@src()` in C (which is something I thought might be a problem for using dvui with C)

```C
#include <stdio.h>
#define SRC {__FILE__, __LINE__}

typedef struct {
    char*file;
    int line;
} Source;

int main() {
    Source a = SRC;
    printf("%s%i", a.file, a.line);
}
```

**Comment by david-vanderson** [comment]
That is very interesting - well done!

**Comment by david-vanderson** [comment]
Cleaning up issues.  Please reopen if needed.
